File: android-beans/src/test/java/rocks/frieler/android/beans/scopes/prototype/PrototypeScopedFactoryBeanHandlerTest.java
package rocks.frieler.android.beans.scopes.prototype;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.CoreMatchers.sameInstance;
import static org.junit.Assert.*;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static rocks.frieler.android.beans.scopes.prototype.PrototypeScopedFactoryBeanHandler.PROTOTYPE_SCOPE;

@RunWith(MockitoJUnitRunner.class)
public class PrototypeScopedFactoryBeanHandlerTest {
    private PrototypeScopedFactoryBeanHandler prototypeScopedFactoryBeanHandler = new PrototypeScopedFactoryBeanHandler();

    @Mock
    private PrototypeScopedFactoryBean<Object> scopedFactoryBean;

    @Test
    public void testGetNameReturnsPrototypeScope() {
        assertThat(prototypeScopedFactoryBeanHandler.getName(), is(PROTOTYPE_SCOPE));
    }

    @Test
    public void testIsAlwaysActive() {
        assertThat(prototypeScopedFactoryBeanHandler.isActive(), is(true));
    }

    @Test
    public void testGetBeanLetsFactoryBeanProduceBean() {
        final Object bean = new Object();
        when(scopedFactoryBean.produceBean()).thenReturn(bean);

        final Object beanInstance = prototypeScopedFactoryBeanHandler.getBean("bean", scopedFactoryBean);

        assertThat(beanInstance, is(sameInstance(bean)));
    }

    @Test
    public void testGetBeanProducesNewBeanEveryTime() {
        final Object firstBean = new Object();
        final Object secondBean = new Object();
        when(scopedFactoryBean.produceBean()).thenReturn(firstBean, secondBean);

        final Object firstBeanInstance = prototypeScopedFactoryBeanHandler.getBean("bean", scopedFactoryBean);
        final Object secondBeanInstance = prototypeScopedFactoryBeanHandler.getBean("bean", scopedFactoryBean);

        verify(scopedFactoryBean, times(2)).produceBean();
        assertThat(firstBeanInstance, is(sameInstance(firstBean)));
        assertThat(secondBeanInstance, is(sameInstance(secondBean)));
    }
}

File: android-beans/src/test/kotlin/rocks/frieler/android/beans/scopes/prototype/PrototypeScopedFactoryBeanHandlerTest.kt
package rocks.frieler.android.beans.scopes.prototype

import assertk.assertThat
import assertk.assertions.isEqualTo
import assertk.assertions.isSameAs
import assertk.assertions.isTrue
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import org.mockito.kotlin.mock
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever
import rocks.frieler.android.beans.BeansProvider

class PrototypeScopedFactoryBeanHandlerTest {

    private val prototypeScopedFactoryBeanHandler = PrototypeScopedFactoryBeanHandler()

    private val scopedFactoryBean: PrototypeScopedFactoryBean<Any> = mock()

    @Test
    fun `name is prototype scope`() {
        assertThat(prototypeScopedFactoryBeanHandler.name).isEqualTo(PrototypeScopedFactoryBeanHandler.PROTOTYPE_SCOPE)
    }

    @Test
    fun `prototype scope is always active`() {
        assertThat(prototypeScopedFactoryBeanHandler.isActive).isTrue()
    }

    @Test
    fun `getBean() lets FactoryBean produce the bean on invocation`() {
        val dependencies: BeansProvider = mock()
        val bean = Any()
        whenever(scopedFactoryBean.produceBean(dependencies)).thenReturn(bean)

        val beanInstance = prototypeScopedFactoryBeanHandler.getBean("bean", scopedFactoryBean, dependencies)

        assertThat(beanInstance).isSameAs(bean)
    }

    @Test
    fun `getBean() produces a new bean instance every time`() {
        val beansProvider: BeansProvider = mock()
        val firstBean = Any()
        val secondBean = Any()
        whenever(scopedFactoryBean.produceBean(beansProvider)).thenReturn(firstBean, secondBean)

        val firstBeanInstance = prototypeScopedFactoryBeanHandler.getBean("bean", scopedFactoryBean, beansProvider)
        val secondBeanInstance = prototypeScopedFactoryBeanHandler.getBean("bean", scopedFactoryBean, beansProvider)

        verify(scopedFactoryBean, Mockito.times(2)).produceBean(beansProvider)
        assertThat(firstBeanInstance).isSameAs(firstBean)
        assertThat(secondBeanInstance).isSameAs(secondBean)
    }
}
